src/Control/Distributed/Process/Internal/Primitives.hs

```diff
@@ -986,7 +986,10 @@ expectTimeout n = receiveTimeout n [match return]
 spawnAsync :: NodeId -> Closure (Process ()) -> Process SpawnRef
 spawnAsync nid proc = do
   spawnRef <- getSpawnRef
-  sendCtrlMsg (Just nid) $ Spawn proc spawnRef
+  node     <- getSelfNode
+  if nid == node
+    then sendCtrlMsg Nothing $! Spawn proc spawnRef
```

Comment:
check again the latest definition of sendCtrlMsg:
https://github.com/tweag/distributed-process/blob/optimization/spawnAsync/src/Control/Distributed/Process/Internal/Messaging.hs#L195